src/parse/asp/builtins.go

```diff
@@ -587,13 +587,49 @@ func strRFind(s *scope, args []pyObject) pyObject {
 
 func strFormat(s *scope, args []pyObject) pyObject {
 	self := string(args[0].(pyString))
-	for k, v := range s.locals {
-		self = strings.ReplaceAll(self, "{"+k+"}", v.String())
-	}
-	for _, arg := range args[1:] {
-		self = strings.Replace(self, "{}", arg.String(), 1)
+	var buf strings.Builder
+	buf.Grow(len(self)) // most reasonable guess available as to how big it might be
+
+	arg := 1 // what arg index are we up to for positional args
+	for {
+		start := strings.IndexByte(self, '{')
+		if start == -1 {
+			buf.WriteString(self)
+			break
+		}
+		end := strings.IndexByte(self[start:], '}')
+		if end == -1 {
+			// We may want to error here in some future revision
+			buf.WriteString(self)
+			break
+		}
+		buf.WriteString(self[:start])
+		end = start + end
+		if start > 0 && self[start-1] == '$' {
+			// Don't interpolate ${X} (but ${{X}} -> ${X})
+			if self[start+1] == '{' {
+				buf.WriteString(self[start+1 : end])
```

Comment:
this could mean
```
${ {X}} -> $ {X}
```
which isn't ideal but unlikely so in the interest of not complicating things more I think its ok especially because in the previous implementation
```
${ {X}} -> ${ {X}
```
which is invalid too

Reply:
Yeah there's a bunch of stuff in there that's a bit gross. This is erring as close as possible to before, but I think we should ban quite a few of these constructions.